Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm

```diff
@@ -0,0 +1,220 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if !__has_feature(objc_arc)
+#error This file requires ARC. Add the "-fobjc-arc" compiler flag for this file.
+#endif
+
+#import "config.h"
+#import "WebExtensionSidebar.h"
+
+#if ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
+
+#import "CocoaHelpers.h"
+#import "WebExtensionContext.h"
+#import "WebExtensionTab.h"
+#import "WebExtensionWindow.h"
+
+namespace WebKit {
+
+static NSString * const fallbackPath = @"about:blank";
+static NSString * const fallbackTitle = @"";
+
+
+static std::optional<String> getDefaultSidebarTitleFromExtension(WebExtension& extension)
+{
+    return toOptional(extension.sidebarTitle());
+}
+
+static std::optional<String> getDefaultSidebarPathFromExtension(WebExtension& extension)
+{
+    return toOptional(extension.sidebarDocumentPath());
+}
+
+static std::optional<NSDictionary *> getDefaultIconsDictFromExtension(WebExtension& extensions)
+{
+    // FIXME: <https://webkit.org/b/276833> implement this
+    return std::nullopt;
+}
+
+WebExtensionSidebar::WebExtensionSidebar(WebExtensionContext& context, IsDefault isDefault) : WebExtensionSidebar(context, std::nullopt, std::nullopt, isDefault) { };
+
+WebExtensionSidebar::WebExtensionSidebar(WebExtensionContext& context, WebExtensionTab& tab) : WebExtensionSidebar(context, tab, std::nullopt, IsDefault::No) { };
+
+WebExtensionSidebar::WebExtensionSidebar(WebExtensionContext& context, WebExtensionWindow& window) : WebExtensionSidebar(context, std::nullopt, window, IsDefault::No) { };
+
+WebExtensionSidebar::WebExtensionSidebar(WebExtensionContext& context, std::optional<Ref<WebExtensionTab>> tab, std::optional<Ref<WebExtensionWindow>> window, IsDefault isDefault)
+    : m_context(WeakRef(context)), m_tab(tab), m_window(window), m_isDefault(isDefault)
```

Comment:
Wouldn't we be forming a reference cycle with `WebExtensionContext` otherwise? Given that `WebExtensionContext` will always hold a reference to the current sidebar object in `m_sidebarTabMap` or `m_sidebarWindowMap` (or `m_defaultSidebar`), I thought we couldn't hold a strong reference to `WebExtensionContext` within `WebExtensionSidebar` without forming a cycle.